### PR TITLE
[SELS-TASK] Create API to update profile details [BE]

### PR DESCRIPTION
**Pull request by jules-lucero**
### Asana Link
https://app.asana.com/0/1203020082973659/1203020083494271/f

### Commands
```
cd backend
npm install
npm install -g sequelize-cli
npm run reset
npm start
```
Send a PUT request to `http://localhost/:YOUR_PORT/users/profile` with a query parameter _user_id_ and a request body
```
Sample Request Body:
{
    "first_name": "New",
    "last_name": "Name",
    "email": "newname@gmail.com",
    "current_password": "pass",
    "new_password": "pass123",
    "avatar_url": "https://joeschmoe.io/api/v1/random"
}
```

### Pre-conditions
Need to use **Postman** or any similar service to test the API

### Expected Output
- [x] You should be able to update a user's profile
- [x] You should be prompted with an error when a user is not found
- [x] You should be prompted with an error when the email is already in use
- [x] You should be prompted with an error when the password is incorrect

### Notes
- Backend Only

### Screenshots
image

### CODE REVIEW

File: backend/controllers/users.js [hunk 1]
```diff
@@ -85,6 +89,63 @@ module.exports = {
...
+    const hash = await bcrypt.hash(new_password, salt);
+
+    user.set({
+      first_name,
+      last_name,
+      email,
+      password: hash,
+      avatar_url,
+    });
+    await user.save();
```
Comment (escbooster12-dev) [inline backend/controllers/users.js:133]: - if it involves a series of queries. It's better to move to model class.
```
const { Model } = require('sequelize');
module.exports = (sequelize, DataTypes) => {
  class User extends Model {
    static associate(models) {
      ...
    }
  }
  User.init(
    ...
  );

  User.getLearntWordsAndLessons = function () {
    console.log("here are the learnt words and lessons");
  };

  return User;
};
```

File: backend/index.js [hunk 2]
```diff
@@ -10,8 +13,34 @@ const wordRoutes = require("./routes/words");
...
+    file.mimetype == "image/png" ||
+    file.mimetype == "image/jpg" ||
+    file.mimetype == "image/jpeg"
+  ) {
+    cb(null, true);
+  } else {
+    cb(null, false);
+    return cb(new Error("Only .png, .jpg and .jpeg format allowed!"));
+  }
+};
```
Comment (escbooster12-dev) [inline backend/index.js:37]: - move these two functions to separate file under `hlepers` directory. Something like `helpers\fileHelper.js`

File: backend/.env.template [hunk 3]
```diff
@@ -1,4 +1,5 @@
 DATABASE_USER=
 DATABASE_PASS=
 DATABASE=
-PORT=
\ No newline at end of file
+PORT=
+BACKEND_URL=
```
Comment (escbooster12-dev) [inline backend/.env.template:5]: - if you have this icon. please add newline

File: backend/controllers/users.js [hunk 4]
```diff
@@ -8,6 +10,8 @@ const {
   Lesson_word,
 } = require("../models");
 
+const salt = 11;
```
Comment (escbooster12-dev) [inline backend/controllers/users.js:13]: - what is this constant variable for?

**Review by escbooster12-dev — CHANGES_REQUESTED** [review]

File: backend/controllers/users.js (hunk 4 above)
Comment (jules-lucero) [reply backend/controllers/users.js:13]: The salt variable is used in bcrypt when hashing a password.

**Review by escbooster12-dev — APPROVED** [review]

**Pull request merged**